### PR TITLE
Add futures01 crate

### PR DESCRIPTION
**Pull request by cramertj**
cc @shepmaster

### CODE REVIEW

File: futures01/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+[package]
+name = "futures01"
+edition = "2018"
```
Comment (taiki-e) [inline futures01/Cargo.toml]: `edition` key is only available starting in Rust 1.31. (AFAIK I think the motivation for this PR was to have 0.1 and 0.3 co-exist in a version below Rust 1.31.)

Comment (cramertj) [reply futures01/Cargo.toml]: *sigh* that... makes sense :)

Comment (cramertj) [reply futures01/Cargo.toml]: cc @shepmaster I'm not sure how to do macro-reexporting on old editions. @dtolnay do you have any tricks here?

**Comment by seanmonstar** [comment]
I *believe* glob imports like you've will include macros.

**Comment by cramertj** [comment]
@seanmonstar that's true in 2018 edition, but not if I change it to 2015 edition :(

**Comment by shepmaster** [comment]
For my own selfish needs, I'm actually OK with the currently-published version, but I agree it's not ideal for other users.

My main problem was that someone using Rust 1.18 but not opting-in to my feature flag for futures would run into a problem because my Cargo.toml would have something like:

```toml
[dependencies]
my_futures_01_name = { package = "futures", version = "0.1", optional = true }
```

Compiling this with Rust 1.18 yields:

```
$ cargo test
    Updating registry `https://github.com/rust-lang/crates.io-index`
error: no matching package named `my_futures_01_name` found (required by `snafu`)
location searched: registry https://github.com/rust-lang/crates.io-index
version required: ^0.1
```

Right now, this doesn't happen because there *is* a crate called `futures01`. If someone tries to compile with my feature flag enabled and an old version of Rust, they will fail. However, I'm moderately OK with forcing people who opt-in to futures needing to have a newer version, although it's not ideal.

**Comment by shepmaster** [comment]
> forcing people who opt-in to futures needing to have a newer version

This would be sad though, since futures 0.1.27 *does* compile on Rust 1.18.

What about a really "exciting" idea:

```toml
[lib]
path = "../real/futures/lib.rs"
```

**Comment by seanmonstar** [comment]
@cramertj ah shucks. I guess, there aren't that many macros in 0.1, just `try_ready` I think, it could just be redefined to do the same thing?

**Comment by cramertj** [comment]
@shepmaster that would compile the lib twice, so the `futures01` versions of traits wouldn't match the `futures` 0.1 ones.

**Comment by cramertj** [comment]
@seanmonstar yeah, `try_ready` and `task_local`. I can just copy them over, though it does seem unfortunate.

**Comment by taiki-e** [comment]
Hmm... The following errors occurred with 1.20 - 1.27:

```text
error: a macro named `try_ready` has already been exported
   --> futures01\src\lib.rs:196:1
    |
196 | / macro_rules! try_ready {
197 | |     ($e:expr) => (match $e {
198 | |         Ok($crate::Async::Ready(t)) => t,
199 | |         Ok($crate::Async::NotReady) => return Ok($crate::Async::NotReady),
200 | |         Err(e) => return Err(From::from(e)),
201 | |     })
202 | | }
    | |_^ `try_ready` already exported
error: a macro named `try_ready` has already been exported
   --> futures01\src\lib.rs:196:1
    |
196 | / macro_rules! try_ready {
197 | |     ($e:expr) => (match $e {
198 | |         Ok($crate::Async::Ready(t)) => t,
199 | |         Ok($crate::Async::NotReady) => return Ok($crate::Async::NotReady),
200 | |         Err(e) => return Err(From::from(e)),
201 | |     })
202 | | }
    | |_^ `try_ready` already exported
    |
    |
note: previous macro export here
   --> futures01\src\lib.rs:164:9
    |
164 | pub use futures::*;
    |         ^^^^^^^^^^

note: previous macro export here
   --> futures01\src\lib.rs:164:9
    |
164 | pub use futures::*;
    |         ^^^^^^^^^^

error: a macro named `task_local` has already been exported
   --> futures01\src\lib.rs:175:1
    |
175 | / macro_rules! task_local {
176 | |     (static $NAME:ident: $t:ty = $e:expr) => (
177 | |         static $NAME: $crate::task::LocalKey<$t> = {
178 | |             fn __init() -> $t { $e }
...   |
188 | |     )
189 | | }
    | |_^ `task_local` already exported
error: a macro named `task_local` has already been exported
   --> futures01\src\lib.rs:175:1
    |
175 | / macro_rules! task_local {
176 | |     (static $NAME:ident: $t:ty = $e:expr) => (
177 | |         static $NAME: $crate::task::LocalKey<$t> = {
178 | |             fn __init() -> $t { $e }
...   |
188 | |     )
189 | | }
    | |_^ `task_local` already exported
    |
    |
note: previous macro export here
   --> futures01\src\lib.rs:164:9
    |
164 | pub use futures::*;
    |         ^^^^^^^^^^

note: previous macro export here
   --> futures01\src\lib.rs:164:9
    |
164 | pub use futures::*;
    |         ^^^^^^^^^^

error: aborting due to 2 previous errors
```

**Comment by shepmaster** [comment]
Ditto with 1.18:

```
$ rustc --version
rustc 1.18.0 (03fc9d622 2017-06-06)

$ cargo test
 Downloading backtrace v0.3.16
   Compiling cc v1.0.37
   Compiling futures v0.1.27 (https://github.com/cramertj/futures-rs/?branch=futures01#6066a033)
   Compiling backtrace v0.3.16
   Compiling snafu-derive v0.3.1 (file:///Users/shep/Projects/snafu/snafu-derive)
   Compiling futures01 v0.1.27 (https://github.com/cramertj/futures-rs/?branch=futures01#6066a033)
error: a macro named `try_ready` has already been exported
   --> /Users/shep/.cargo/git/checkouts/futures-rs-3cb6df2ed2f884c5/6066a03/futures01/src/lib.rs:196:1
    |
196 | / macro_rules! try_ready {
197 | |     ($e:expr) => (match $e {
198 | |         Ok($crate::Async::Ready(t)) => t,
199 | |         Ok($crate::Async::NotReady) => return Ok($crate::Async::NotReady),
200 | |         Err(e) => return Err(From::from(e)),
201 | |     })
202 | | }
    | |_^ `try_ready` already exported
    |
note: previous macro export here
   --> /Users/shep/.cargo/git/checkouts/futures-rs-3cb6df2ed2f884c5/6066a03/futures01/src/lib.rs:164:9
    |
164 | pub use futures::*;
    |         ^^^^^^^^^^^
```

Indeed, commenting out those two macros seems to let my code compile. Was there just a bug in macros that we can make use of, or does it turn out none of us know how macros work? 😇

File: futures01/Cargo.toml (hunk 1 above)
Comment (dtolnay) [reply futures01/Cargo.toml]: A glob re-export (`pub use dependency::*`) will include macros back to rustc 1.15.0 on any edition.

Comment (cramertj) [reply futures01/Cargo.toml]: Ah, wonderful-- thank you!

Comment (cramertj) [reply futures01/Cargo.toml]: (seems incredibly surprising to me that `use dependency::*;` wouldn't actually bring the macro into scope, but `pub use dependency::*;` would reexport it :/ )

Comment (dtolnay) [reply futures01/Cargo.toml]: Indeed. IIRC it was stabilized by accident.

**Comment by shepmaster** [comment]
0737933281f82ea7455cd0aca4cc12defa44835e appears to compile for me in Rust 1.18; thank you!

**Pull request merged**